### DISCUSSION TITLE
[QUESTION]

### DISCUSSION THREAD

**Discussion by dann1kid** [python-telegram-bot/python-telegram-bot#2385] [Q&A]
### Issue I am facing
I can't route callbacks from inline to multiple ConversationHandlers .
If i add another one route to another Conversation handler, another callback routes to the 1st ConversationHandler.
If it one it work correctly with one type of button (take payment), but when i add and click on 'take comment' it react like i push 1st button...
Any tips or code snippets please...

### Code is simiplificated

main 
```
from get_payments_conv_handler import get_payments_on_deliver_handler
from save_comment_conv_handler import  save_users_status_in_work_handler

self.dispatcher.add_handler(get_payments_on_deliver_handler)                 #1st route
self.dispatcher.add_handler(save_users_status_in_work_handler)                #2nd route
```

### 1st ConversationHandler
```
""" starts Conversation to fill  payments """

entry_point_pattern = "^take payments"

def take_payment(update, context):
    pass

def save_payments(update, context):
    pass


get_payments_on_deliver_handler = ConversationHandler(
    entry_points=[CallbackQueryHandler(take_payment, entry_point_pattern)],

    states={
        TAKE_PRICE: [MessageHandler(Filters.text, save_payments)],
    },
    fallbacks=[CommandHandler('cancel', cancel)], allow_reentry=True,
)
```

keyboard with callback button for this conversation

```
keyboard = [
        first line of buttons
        [
            InlineKeyboardButton(f"❌ TAKE PAYMENTS! ❌",
                                 callback_data=f"take payments {order_id} {delivery_id}"),
        ],

        end of buttons
    ]
    return InlineKeyboardMarkup(keyboard)
```

### 2nd ConversationHandler
```
""" starts Conversation to fill comments"""

entry_point_pattern = "^take comments"


def take_comment(update, context):
    pass

def save_comment(update, context):
    pass

save_users_status_in_work_handler = ConversationHandler(
    entry_points=[CallbackQueryHandler(take_comment, entry_point_pattern)],

    states={
        TAKE_COMMENT: [MessageHandler(Filters.text, save_comment)],
    },
    fallbacks=[CommandHandler('cancel', cancel)], allow_reentry=True,
)
```

keyboard with callback button for this conversation
```
keyboard = [
        first line of buttons
        [
            InlineKeyboardButton(f"Add comment to this order", callback_data=f"take comments {order_id} {work_id}"),  
            # here f strings with chained 
            # data like order_id from db
        ],

        end of buttons
    ]
    return InlineKeyboardMarkup(keyboard)
```

**Comment by bqback** [comment]
Would I be right to assume you're in a conversation already and both of these conversation handlers are second (or deeper) level conversations?

**Reply by dann1kid** [reply]
Okay, how I can start deeper convesations in this case?

**Reply by bqback** [reply]
See [this](https://github.com/python-telegram-bot/python-telegram-bot/blob/master/examples/nestedconversationbot.py) example. If your `ConversationHandler` is part of an existing conversation, you don't need to register it separately, just list it as one of the handlers in the state you need it in. If you want another example, check [this](https://github.com/bqback/GKB64-Survey-Bot/blob/main/bot/setup.py) file, where `add_survey` is defined on [line 91](https://github.com/bqback/GKB64-Survey-Bot/blob/main/bot/setup.py#L91) and added as a second level conversation on [line 164](https://github.com/bqback/GKB64-Survey-Bot/blob/main/bot/setup.py#L164). Same for `edit_survey`, a third level `ConversationHandler`, defined on [line 45](https://github.com/bqback/GKB64-Survey-Bot/blob/main/bot/setup.py#L45) and added as a third level conversation on [line 141](https://github.com/bqback/GKB64-Survey-Bot/blob/main/bot/setup.py#L141). Neither are registered with `add_handler`, but they work as intended.

**Reply by dann1kid** [reply]
Alas, this option is not very suitable, because it is built with the possibility of embedding the ConversionHandler on the callback. If you introduce parent handlers, then it will be cumbersome and not modular, I thought about such an implementation, but then it will take a lot of time to implement the functionality that already exists.

I think this is either the wrong callback fade filtering for this case, or it's an API error
